pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,580 @@
+
+		jobListerSynced:     jobInformer.Informer().HasSynced,
+		cronJobListerSynced: cronJobsInformer.Informer().HasSynced,
+		now:                 time.Now,
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    jm.addJob,
+		UpdateFunc: jm.updateJob,
+		DeleteFunc: jm.deleteJob,
+	})
+
+	cronJobsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+		UpdateFunc: jm.updateCronJob,
+		DeleteFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+	})
+
+	return jm, nil
+}
+
+// Run starts the main goroutine responsible for watching and syncing jobs.
+func (jm *Controller2) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	klog.Infof("Starting cronjob controller")
+
+	if !cache.WaitForNamedCacheSync("job", stopCh, jm.jobListerSynced, jm.cronJobListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(jm.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	klog.Infof("Shutting down CronJob Manager")
+}
+
+func (jm *Controller2) worker() {
+	for jm.processNextWorkItem() {
+	}
+}
+
+func (jm *Controller2) processNextWorkItem() bool {
+	key, quit := jm.queue.Get()
+	if quit {
+		return false
+	}
+	defer jm.queue.Done(key)
+	err, requeueAfter := jm.sync(key.(string))
+	switch {
+	case err != nil:
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		jm.queue.Add(key)
+	case requeueAfter != nil:
+		jm.queue.AddAfter(key, *requeueAfter)
+	default:
+		jm.queue.Done(key)
+	}
+	return true
+}
+
+func (jm *Controller2) sync(cronJobKey string) (error, *time.Duration) {
+	ns, name, err := cache.SplitMetaNamespaceKey(cronJobKey)
+	if err != nil {
+		return err, nil
+	}
+
+	cronJob, err := jm.cronJobLister.CronJobs(ns).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// may be cronjob is deleted, dont need to requeue this key
+		nameForLog := fmt.Sprintf("%s/%s", ns, name)
+		klog.V(4).Infof("cronjob %s not found, may be it is deleted: %#v", nameForLog, err)
+		return nil, nil
+	case err != nil:
+		// for other transient apiserver error requeue with exponential backoff
+		return err, nil
+	}
+
+	// TODO: @alpatel limit the amount of data we scrape here
+	jobList, err := jm.jobLister.Jobs(ns).List(labels.Everything())
+	if err != nil {
+		return err, nil
+	}
+
+	jobsToBeReconciled := []batchv1.Job{}
+
+	for _, job := range jobList {
+		// If it has a ControllerRef, that's all that matters.
+		if controllerRef := metav1.GetControllerOf(job); controllerRef != nil && controllerRef.Name == name {
+			// this job is needs to be reconciled
+			jobsToBeReconciled = append(jobsToBeReconciled, *job)
+		}
+	}
+
+	err, requeueAfter := syncOne2(cronJob, jobsToBeReconciled, time.Now(), jm.jobControl, jm.cronJobControl, jm.recorder)
+	// TODO: @alpatel need to rework this function to handle re-queuing on errors
```

Comment:
You don't re-queue here. Return error and `processNextWorkItem` handles re-queuing.

Reply:
+1 I agree, that todo was just to remind me to refactor the function before we make a final first draft, looks like it worked, I'll refactor the function now, thanks